### DISCUSSION TITLE
Convert Avatar's AP from scalar value in its state to FungibleAssetValue

### DISCUSSION THREAD

**Discussion by longfin** [planetarium/lib9c#1293] [Ideas]
Note: For sake of comms, this discussion and its comments will be written in Korean, and may contains private links from team Notion.🙏

---

나인 크로니클의 블록 거버넌스 문제를 해결하기 위해 작업 증명(Proof of Work, 이하 PoW)에서 지분 증명(Proof of Stake, 이하 PoS)를 도입 중에 있으나, PoS만으로는 이 문제를 온전히 해결하는 것이 불가능합니다. 왜냐하면, 나인 크로니클의 트랜잭션들은 수수료를 포함하지 않으며, 이로 인해 마이너(in PoW)나 벨리데이터(in PoS)는 트랜잭션을 자신의 블록에 포함할 유인이 없다는 근본적인 문제는 해결되지 않았기 때문입니다. 이를 해결하기 위해 [AP 순환 시스템]이 제안 되었지만, 이에 관련한 개발은 v100280 까진 이뤄지지 않은 상태입니다. 

이 문제를 보다 쉽게 다루기 위해서, 본격적인 시스템 구현 전에 [`Avatar.actionPoint`](https://github.com/planetarium/lib9c/blob/v100260/Lib9c/Model/State/AvatarState.cs#L40)로 다루고 있는 AP를 Libplanet 에서 제공하는 [`FungibleAssetValue`]로 바꾸는 것을 제안합니다.

구체적인 실행 안은 관련 작업을 많이 하신 @ipdae @boscohyun @hskim1028 님의 의견을 듣고 정하려고 합니다.

[AP 순환 시스템]: https://www.notion.so/planetarium/AP-5f5e62a6a6864d5d9b7dc8fa434eaec9
[`FungibleAssetValue`]: https://docs.libplanet.io/0.40.0/api/Libplanet.Assets.FungibleAssetValue.html

**Comment by riemannulus** [comment]
More information: https://www.notion.so/planetarium/Nine-Chronicles-PoS-token-economy-proposal-b1a47e6b8d20478d99bd062f33a976ae

**Reply by riemannulus** [reply]
짧은 요약:
- AP를 거버넌스 참여 보상으로 주기 위해서는 libplanet에서 AP를 어떠한 타입으로든 다룰 수 있어야 함.
- 하지만 현재 AP는 lib9c에서 순수하게 상태로만 구현되어 있기 때문에, libplanet 에서 관리할 수 있는 방법이 없음.
- 따라서 AP를 libplanet에서 다룰 수 있게 새로운 타입을 만들기보다, AP를 FAV로 변환함으로써 libplanet에서 토큰과 동일하게 다룰 수 있게 만들어야 함.

**Comment by ipdae** [comment]
사전준비라고 되있는데 `FungibleAssetValue` 로 변경하는건 상태를 확인해볼 필요없게 하려고 하는건가요?

**Reply by longfin** [reply]
> 사전준비라고 되있는데 `FungibleAssetValue` 로 변경하는건 상태를 확인해볼 필요없게 하려고 하는건가요?

네. 근데 아주 정확하게는, 상태를 조회할 필요가 없는 점도 있긴 한데, Lib9c가 아닌 Libplanet에서 app agnostic 하게 다룰 수 있어야 해서 (그래야 수수료 시스템이 설계 가능하니...) FAV로 바꾸려는 의도입니다.

**Comment by ipdae** [comment]
`FungibleAssetValue` 로 변경하고 확인을 하는 부분들은 크게 문제가 없을것같은데 AP를 리필할때(`DailyReward, ChargeActionPoint`)  `FungibleAssetValue.Mint` 를 호출하게 되면 이 부분이 문제가 없을지 확인해야할 것 같습니다.

**Reply by longfin** [reply]
> `FungibleAssetValue` 로 변경하고 확인을 하는 부분들은 크게 문제가 없을것같은데 AP를 리필할때(`DailyReward, ChargeActionPoint`) `FungibleAssetValue.Mint` 를 호출하게 되면 이 부분이 문제가 없을지 확인해야할 것 같습니다.

NCG나 다른 일반적인 토큰들처럼 직접적으로 총량을 제한하기 어렵기도 할 것 같고 그래야 할 동인도 좀 떨어져서, FAV화 하더라도 발행 (mint) / 소각(burn)에 대해서는 특별히 제한을 두지 않는 걸 생각하고 있습니다. (@planetarium/libplanet 혹시 문제가 될까요?)

**Reply by riemannulus** [reply]
NCG도 제한을 두더라도 lib9c에서 둔 것이지, libplanet 자체에서 둔 건 아니라 비슷한 맥락으로 그 부분은 별도로 다뤄야 할 이야기 (본격적인 [AP 순환 시스템](https://www.notion.so/planetarium/AP-5f5e62a6a6864d5d9b7dc8fa434eaec9) 설계 시점) 라고 생각합니다.
